pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -108,30 +117,31 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(env *model.Environme
 			}
 		}
 
-		pluginParams := &plugin.InputParams{
-			ListenerProtocol: listenerType,
-			ListenerCategory: networking.EnvoyFilter_ListenerMatch_GATEWAY,
-			Env:              env,
-			Node:             node,
-			ProxyInstances:   workloadInstances,
-			Push:             push,
-			ServiceInstance:  si,
-			Port: &model.Port{
-				Name:     servers[0].Port.Name,
-				Port:     int(portNumber),
-				Protocol: protocol,
-			},
-		}
-		for _, p := range configgen.Plugins {
-			if err := p.OnOutboundListener(pluginParams, mutable); err != nil {
-				log.Warna("buildGatewayListeners: failed to build listener for gateway: ", err.Error())
+		for listenerType := range serversByProtocol {
+			pluginParams := &plugin.InputParams{
+				ListenerProtocol: listenerType,
+				ListenerCategory: networking.EnvoyFilter_ListenerMatch_GATEWAY,
+				Env:              env,
+				Node:             node,
+				ProxyInstances:   workloadInstances,
+				Push:             push,
+				ServiceInstance:  si,
+				Port: &model.Port{
+					Name: servers[0].Port.Name,
+					Port: int(portNumber),
+				},
+			}
+			for _, p := range configgen.Plugins {
+				if err := p.OnOutboundListener(pluginParams, mutable); err != nil {
+					log.Warna("buildGatewayListeners: failed to build listener for gateway: ", err.Error())
+				}
 			}
-		}
 
-		// Filters are serialized one time into an opaque struct once we have the complete list.
-		if err := buildCompleteFilterChain(pluginParams, mutable, opts); err != nil {
-			errs = multierror.Append(errs, fmt.Errorf("gateway omitting listener %q due to: %v", mutable.Listener.Name, err.Error()))
-			continue
+			// Filters are serialized one time into an opaque struct once we have the complete list.
+			if err := buildCompleteFilterChain(pluginParams, mutable, opts); err != nil {
```

Comment:
I am not sure whether this is right for two different listener protocol types build filterChain twice